test/unit/Cardano/WalletLayerSpec.hs

```diff
@@ -0,0 +1,127 @@
+    ) where
+
+
+import Prelude
+
+import Cardano.DBLayer
+    ( DBLayer (..), PrimaryKey (..) )
+import Cardano.DBLayer.MVar
+    ( newDBLayer )
+import Cardano.NetworkLayer.HttpBridgeSpec
+    ( mockNetworkLayer, noLog )
+import Cardano.Wallet
+    ( WalletId (..), WalletName (..) )
+import Cardano.Wallet.AddressDerivationSpec
+    ()
+import Cardano.Wallet.AddressDiscovery
+    ( SeqState )
+import Cardano.Wallet.AddressDiscoverySpec
+    ()
+import Cardano.Wallet.MnemonicSpec
+    ()
+import Cardano.Wallet.Primitive
+    ( SlotId (..) )
+import Cardano.WalletLayer
+    ( NewWallet (..), WalletLayer (..), mkWalletLayer )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( runExceptT )
+import Data.Either
+    ( isLeft, isRight )
+import Data.Maybe
+    ( isJust )
+import Test.Hspec
+    ( Spec, describe, it, shouldSatisfy )
+import Test.QuickCheck
+    ( Arbitrary (..), Property, checkCoverage )
+import Test.QuickCheck.Monadic
+    ( monadicIO )
+
+import qualified Data.Text as T
+
+spec :: Spec
+spec = do
+    describe "WalletLayer works as expected" $ do
+        it "Wallet upon creation is written down in db and cannot be created more than once"
+            (checkCoverage walletCreationProp)
+        it "Wallet after being created can be got using valid wallet Id"
+            (checkCoverage walletGetProp)
+        it "Wallet with wrong wallet Id cannot be got"
+            (checkCoverage walletGetWrongIdProp)
+
+
+{-------------------------------------------------------------------------------
+                                    Test Logic
+-------------------------------------------------------------------------------}
+
+data WalletLayerFixture = WalletLayerFixture {
+      _fixtureDBLayer :: DBLayer IO SeqState
+    , _fixtureWalletLayer :: WalletLayer IO SeqState
+    }
+
+prepareWalletLayerFixture :: IO WalletLayerFixture
+prepareWalletLayerFixture = do
+    db <- newDBLayer
+    let network = mockNetworkLayer noLog 105 (SlotId 106 1492)
+    let walletLayer = mkWalletLayer db network
+    return $ WalletLayerFixture db walletLayer
+
+
+instance Arbitrary NewWallet where
+    -- No shrinking
+    arbitrary = NewWallet
+                <$> arbitrary
+                <*> arbitrary
+                <*> pure (WalletName "My Wallet")
+                <*> arbitrary
+                <*> arbitrary
+
+
+walletCreationProp
+    :: NewWallet
+    -> Property
+walletCreationProp newWallet = monadicIO $ liftIO $ do
+    (WalletLayerFixture db wl) <- liftIO prepareWalletLayerFixture
+    res <- runExceptT $ createWallet wl newWallet
+    case res of
+        Left e ->
+            fail $ show e
+        Right walletId -> do
+            resFromDb <- readCheckpoints db (PrimaryKey walletId)
+            resFromDb `shouldSatisfy` isJust
+            --the same wallet creation second time
+            secondTrial <- runExceptT $ createWallet wl newWallet
+            secondTrial `shouldSatisfy` isLeft
```

Comment:
I'd advise to make two separate tests for this. One test that focuses on creating wallets with various parameters from `NewWallet`, and one test that specifically test that a wallet can't be inserted twice. It's good to keep tests focused to one and one thing such that, when a test fail, you know exactly why it fails.

Reply:
now, `setupFixture` prepares db, walletLayer and creates wallet giving its walletId

Reply:
seperated this test into two as suggested